### ISSUE TITLE
Integration testing

### ISSUE THREAD

**Issue by dead10ck** [spicavigo/kafka-rust#138]
Currently, much of the code that requires doing actual networking calls to a Kafka server are not tested. We can take advantage of Docker and Travis to enable integration testing the networking methods. I did some research and playing around today, and I came up with a setup that seems to work.

Basically, we set up a `docker-compose.yml` that uses a public Kafka container from the Docker registry, and give the Kafka version as a parameter. Then we set up our `.travis.yml` to do one test for each Kafka version we want to test against. This will allow us not only to ensure sure the code works as intended, but also verify compatibility across as many versions of Kafka as we'd like.

[This commit](https://github.com/dead10ck/kafka-rust/commit/25633eb0b5ab4a78732e20c0c726562f3a763d8b) in a branch of my fork showcases the code changes necessary. It uses the latest versions of 0.8, 0.9, and 0.10, and includes a simple integration test for loading the `KafkaClient` metadata.

What do you think?

**Comment by xitep** [comment]
yes, that definitely is quite cool! 👍 (many many thanks for the initiative!) it's a start to build up a integration test suite - which is missing so far. Some points/ideas coming to my mind:

* can we run `cargo test` without the integration tests? (executing merely the unit tests; i'd say during development one wants to run the unit tests more often than the full integration tests since the latter typically run longer.)
* i think it would be good to have the possibility to run all of this - maybe with a single script - on a local machine without the dependence on travis (assuming docker is available and running on the machine.)
* i was wondering: how comes the test in your commit can assert exactly a single topic in the kafka broker with exactly two partitions? where is it defined that such a topic - and only this one - exists in the broker?
* since `kafka-rust` is not covering any of the kafka administrative APIs, we might need to further think about how to reset the state in kafka between individual integration tests.
* further, i believe it would be great trying to make a set up which includes more than one broker. there's lots of potentials in the code for errors when considering a cluster instead of a single broker kafka set-up.

I hope these ideas can be helpful.

**Comment by dead10ck** [comment]
> can we run `cargo test` without the integration tests?

Yep, `cargo test` has the `--doc`, `--lib`, and `--test` options for running the documentation, unit, and integration tests, respectively.

> i think it would be good to have the possibility to run all of this - maybe with a single script - on a local machine without the dependence on travis

Yep, as I described, all you need to do is set the `KAFKA_VER` env variable, and then do `docker-compose up -d`, and then you're good to go. It would be simple to write a script or a Makefile to do this automatically.

> i was wondering: how comes the test in your commit can assert exactly a single topic in the kafka broker with exactly two partitions? where is it defined that such a topic - and only this one - exists in the broker?

Sorry, this was not obvious. The `wurstmeiser/kafka` Docker container is configured via its environment variables; the automatic creation of topics is controlled by the [`KAFKA_CREATE_TOPICS` variable set here](https://github.com/dead10ck/kafka-rust/commit/25633eb0b5ab4a78732e20c0c726562f3a763d8b#diff-4e5e90c6228fd48698d074241c2ba760R18).

> since kafka-rust is not covering any of the kafka administrative APIs, we might need to further think about how to reset the state in kafka between individual integration tests.

Yes, we'll definitely have to think about how to control broker state, because the larger the test suite gets, the more they will get tangled with each other. It may even be necessary to not run the integration tests in parallel.

> further, i believe it would be great trying to make a set up which includes more than one broker. there's lots of potentials in the code for errors when considering a cluster instead of a single broker kafka set-up.

Indeed, we will definitely need to test clustered environments as well; the `wurstmeister/kafka` container does have the ability to scale to multiple nodes, but I haven't played with that yet.

We can use this issue as an ongoing place for discussion of these problems, until they're mostly ironed out.

**Comment by dead10ck** [comment]
> > i think it would be good to have the possibility to run all of this - maybe with a single script - on a local machine without the dependence on travis
> 
> Yep, as I described, all you need to do is set the KAFKA_VER env variable, and then do docker-compose up -d, and then you're good to go. It would be simple to write a script or a Makefile to do this automatically.

I just added a script named `test` to my branch which takes care of the docker setup and runs all tests with all Kafka versions automatically, if you'd like to check it out.

**Comment by xitep** [comment]
@dead10ck thanks for all your answers and the script. the idea becomes clearer to me. i like it! 👍 i'm a bit worried about the fact that the we'd need to spread the the topic creation (i.e. setting up the test) from the actual tests (this might lead to a difficult-to-maintain test-suite.) currently, this is our only chance. and as you'd said we can improve on this in the future.

**Comment by dead10ck** [comment]
If you're a newcomer, welcome! I have a PR open that adds a basic integration testing framework via Docker: #143. If you're familiar with Docker setups, I would appreciate a review! The tests in the PR are rather basic, but I wanted to get the framework settled first. Once it's merged, others can begin backfilling integration tests.

**Comment by dead10ck** [comment]
This has been merged in #143. Let's keep this issue around to track backfilling integration tests for the rest of the crate, where appropriate.

**Comment by AlexandreCassagne** [comment]
Is this issue still relevant?
